services/src/main/scala/cromwell/services/metadata/impl/WriteMetadataActor.scala

```diff
@@ -38,15 +38,36 @@ class WriteMetadataActor(override val batchSize: Int,
       case Success(_) =>
         putWithResponse foreach { case (ev, replyTo) => replyTo ! MetadataWriteSuccess(ev) }
       case Failure(regerts) =>
-        val workflowMetadataFailureCounts = e.toVector.flatMap(_.events).groupBy(x => x.key.workflowId).map { case (wfid, list) => s"$wfid: ${list.size}" }
-        log.error("Metadata events permanently dropped for the following workflows: " + workflowMetadataFailureCounts.mkString(","))
 
-        putWithResponse foreach { case (ev, replyTo) => replyTo ! MetadataWriteFailure(regerts, ev) }
+        val (outOfLives, stillGood) = e.toVector.partition(_.maxAttempts <= 1)
+
+        handleOutOfLives(outOfLives, regerts)
+        handleEventsToReconsider(stillGood)
     }
 
     dbAction.map(_ => allPutEvents.size)
   }
 
+  private def handleOutOfLives(writeActions: Vector[MetadataWriteAction], reason: Throwable): Unit = if (writeActions.nonEmpty) {
+    val workflowMetadataFailureCounts = writeActions.toVector.flatMap(_.events).groupBy(x => x.key.workflowId).map { case (wfid, list) => s"$wfid: ${list.size}" }
+    log.error(reason, "Metadata event writes have failed irretrievably for the following workflows. They will be lost: " + workflowMetadataFailureCounts.mkString(","))
+
+    writeActions foreach {
+      case PutMetadataActionAndRespond(ev, replyTo, _) => replyTo ! MetadataWriteFailure(reason, ev)
+      case _: PutMetadataAction => () // We need to satisfy the exhaustive match but there's nothing special to do here
+    }
+  }
+
+  private def handleEventsToReconsider(writeActions: Vector[MetadataWriteAction]): Unit = if (writeActions.nonEmpty) {
+    val workflowMetadataFailureCounts = writeActions.toVector.flatMap(_.events).groupBy(x => x.key.workflowId).map { case (wfid, list) => s"$wfid: ${list.size}" }
+    log.warning("Metadata event writes have failed for the following workflows. They will be reconsidered: " + workflowMetadataFailureCounts.mkString(","))
+
+    writeActions foreach {
+      case action: PutMetadataAction => self ! action.copy(maxAttempts = action.maxAttempts - 1)
+      case action: PutMetadataActionAndRespond => self ! action.copy(maxAttempts = action.maxAttempts - 1)
+    }
```

Comment:
I've extracted out the log message generation which was pretty much the only copy/pasted section of these two `handle` methods. Otherwise they're actually fairly distinct.

Reply:
Oh, I see you mean the actual `self ! x.copy(...)`...

Yeah, it's a little annoying... the only obvious "fix" occurring to me would be to make a trait-level function and then have the implementations under the two case classes. The trouble is that doesn't feel to me too much like "fixing" this rather than just moving the copy/paste elsewhere.